Modify default 'fallthrough' behavior.

bryce-anderson opened the pull request:
I believe this addresses #416. It attempts to address the shortcomings of PR #418.

kryptt commented on server/src/main/scala/org/http4s/server/Fallthrough.scala:
```diff
@@ -27,7 +27,8 @@ object Fallthrough {
   implicit def forResponse: Fallthrough[Response] = new Fallthrough[Response] {
-    def isFallthrough(r: Response): Boolean = r.status.code == 404
+    def isFallthrough(r: Response): Boolean =
+      r.status.code == 404 && r.attributes.contains(HttpService.notFoundServiceKey)
```
Why not simply:
```scala
def isFallthrough(r: Response): Boolean =
      r.attributes.contains(HttpService.fallthroughServiceKey)
```

bryce-anderson replied:
Because a map lookup is quite costly compared to an integer comparison. I'm
not totally against it but I see it as the purview of an alternative
`Fallthrough`.

On Wed, Oct 7, 2015 at 10:32 AM, Rodolfo Hansen <notifications@github.com>
wrote:

> In server/src/main/scala/org/http4s/server/Fallthrough.scala
> <https://github.com/http4s/http4s/pull/425#discussion_r41397032>:
>
> > @@ -27,7 +27,8 @@ object Fallthrough {
> >    /** A [[Response]] specific [[Fallthrough]] which considers any response with a 404
> >      * status code as a fallthrough. */
> >    implicit def forResponse: Fallthrough[Response] = new Fallthrough[Response] {
> > -    def isFallthrough(r: Response): Boolean = r.status.code == 404
> > +    def isFallthrough(r: Response): Boolean =
> > +      r.status.code == 404 && r.attributes.contains(HttpService.notFoundServiceKey)
>
> Why not simply:
>
> def isFallthrough(r: Response): Boolean =
>       r.attributes.contains(HttpService.fallthroughServiceKey)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/http4s/http4s/pull/425/files#r41397032>.
>

kryptt replied:
Hmmm, I see your point. Agreed then!

I do prefer renaming the attribute key to something that is specific to fallthrough though.

bryce-anderson replied:
That does seem like a good idea. I'll rename it to `fallthroughServiceKey` and merge this.

kryptt commented:
:+1: here (the tests look good too)

bryce-anderson commented:
Thanks for the review!

bryce-anderson commented:
@kryptt, I took your advice a step further and renamed the default fallthrough instance to to make it clear its based on 404 and moved the fallthrough attribute tag to the `Fallthrough` object. Let me know sooner than later if you object: otherwise I'll merge this in a few hours.

kryptt commented:
Looks great!
